certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,215 @@
+// - Safe contract should be the sender of the transaction.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold) {
+    requireSocialRecoveryModuleEnabled();
+
+    require e.msg.value == 0;
+    require threshold > 0;
+    require guardianStorageContract.entries[safeContract].count < max_uint256; // To prevent overflow (Realistically can't reach).
+    requireGuardiansLinkedListIntegrity();
+
+    require guardian != 0;
+    require guardian != SENTINEL();
+    require guardian != safeContract;
+
+    require !safeContract.isOwner(guardian);
+    require !currentContract.isGuardian(safeContract, guardian);
+
+    require threshold <= guardianStorageContract.countGuardians(safeContract);
+
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted && currentContract.isGuardian(safeContract, guardian);
+}
+
+// This integrity rule verifies that the addition of a new guardian always reverts if the guardian is already added.
+rule addGuardianRevertIfDuplicateGuardian(env e, address guardian, uint256 threshold) {
+    bool isGuardian = currentContract.isGuardian(safeContract, guardian);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+
+    assert isGuardian => lastReverted;
+}
+
+// This integrity rule verifies that the addition of a new guardian can revert due to different reasons even if the address is not a guardian.
+rule addGuardianCanRevertEvenIfAddressIsNotGuardian(env e, address otherAccount, uint256 threshold) {
+    bool isNotGuardian = !currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, otherAccount, threshold);
+
+    assert isNotGuardian && lastReverted =>
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        otherAccount == 0 ||
+        otherAccount == SENTINEL() ||
+        otherAccount == safeContract ||
+        safeContract.isOwner(otherAccount) ||
+        threshold == 0 ||
+        threshold >= guardianStorageContract.entries[safeContract].count ||
+        guardianStorageContract.entries[safeContract].count == max_uint256 ||
+        !safeContract.isModuleEnabled(currentContract);
+}
+
+// This integrity rule verifies that if the revokeGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the guardian is revoked from the guardian list
+// - the linked list integrity remains,
+// - and no other account (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+
+    require guardian != otherAccount;
```

Comment:
Same comment about removing the `require`.

Reply:
Done 👍🏾